lib/codegen/src/nan_canonicalization.rs

```diff
@@ -0,0 +1,110 @@
+            }
+        }
+    }
+}
+
+/// Returns true/false based on whether the instruction is a floating-point
+/// arithmetic operation. This ignores operations like `fneg`, `fabs`, or
+/// `fcopysign` that only operate on the sign bit of a floating point value.
+fn is_fp_arith(pos: &mut FuncCursor, inst: Inst) -> bool {
+    // FIXUP: Should `ceil`, `floor`, `trunc`, `nearest`, and
+    // immediate constants be considered as well?
+    match pos.func.dfg[inst] {
+        InstructionData::Unary { opcode, .. } => opcode == Opcode::Sqrt,
+        InstructionData::Binary { opcode, .. } => {
+            opcode == Opcode::Fadd || opcode == Opcode::Fsub || opcode == Opcode::Fmul ||
+                opcode == Opcode::Fdiv || opcode == Opcode::Fmin ||
+                opcode == Opcode::Fmax
+        }
+        InstructionData::Ternary { opcode, .. } => opcode == Opcode::Fma,
+        _ => false,
+    }
+}
+
+/// Patch instructions that may result in a NaN result with operations to
+/// identify and replace NaN's with a single canonical NaN value.
+fn add_nan_canon_instrs(pos: &mut FuncCursor, inst: Inst) {
+    // FIXUP: Are these unwraps safe? An arithmetic operation should never be
+    // found at the end of an EBB, if I understand this correctly.
+
+    // Select the result of the instruction, move to the next instruction.
+    let inst_res: Value = pos.func.dfg.first_result(inst);
+    let next_inst: Inst = pos.next_inst().unwrap();
+
+    // Insert a comparison function, and a canonical NaN constant. Select
+    // the constant value, and move forward to the next instruction.
+    let is_nan: Value = pos.ins().fcmp(FloatCC::NotEqual, inst_res, inst_res);
+    insert_nan_const(pos, inst);
+    let canon_nan_instr: Inst = pos.prev_inst().unwrap();
+    let canon_nan_res: Value = pos.func.dfg.first_result(canon_nan_instr);
+    pos.goto_inst(next_inst);
+
+    // Insert a select instruction to canonicalize the NaN value.
+    pos.ins().select(is_nan, canon_nan_res, inst_res);
+
+    // Move backwards one instruction, so that the loop in the
+    // `do_nan_canonicalization` function does not skip any instructions.
+    pos.prev_inst();
+}
+
+/// Insert the canonical 32-bit or 64-bit NaN constant value at the current
+/// position.
+fn insert_nan_const(pos: &mut FuncCursor, inst: Inst) {
+    match get_nan_type(&pos.func.dfg, inst) {
+        types::F32 => {
+            let canon_nan = Ieee32::with_bits(CANON_32BIT_NAN);
+            pos.ins().f32const(canon_nan);
+        }
+        types::F64 => {
+            let canon_nan = Ieee64::with_bits(CANON_64BIT_NAN);
+            pos.ins().f64const(canon_nan);
+        }
+        _ => {} // FIXUP: Should this panic or throw some sort of Error?
+    }
+}
+
+/// Given some instruction that could potentially return a nondeterministic
+/// NaN value, determine if the operation is using 32-bit or 64-bit floating
+/// point numbers, and return the corresponding NaN value.
```

Comment:
You can get this type from the instruction's result type, rather than digging into the operands. `func.dfg.value_type(func.dfg.first_result(inst))` or so. Which is simpler, but also, that's the value that you'll be canonicalizing.